### DISCUSSION TITLE
After updating from Filament v1.39 to 1.40, the display is very slow

### DISCUSSION THREAD

**Discussion by cx20** [google/filament#6924] [Q&A] [answered]
I've changed the sample library that displays the Duck model from Filament `1.39` to `1.40`.
As a result, it was very late. It may be a GPU-dependent problem, but is there any way to deal with it?

Below is a link to a sample.
Filament `v1.39`
https://rawcdn.githack.com/cx20/gltf-test/67f784ad6db9c48eebb60bd7a7b51c7a43080a98/examples/filament/index.html?category=sampleModels&model=Duck&scale=1&type=glTF

Filament `v1.40`
https://rawcdn.githack.com/cx20/gltf-test/a0c2d78902e20cfbbc68e84c8e779e19ca5797d2/examples/filament/index.html?category=sampleModels&model=Duck&scale=1&type=glTF

In Filament `v1.40`, the following warning is displayed and drawing becomes very slow.

```
index.html:1 [.WebGL-00007B4C00B5F800]GL Driver Message (OpenGL, Performance, GL_CLOSE_PATH_NV, Medium): You could get a better shader compiling performance if you re-write the uniform block(s)
[ FrameUniforms ]
in the FRAGMENT shader.
You could get more details from https://chromium.googlesource.com/angle/angle/+/refs/heads/main/src/libANGLE/renderer/d3d/d3d11/UniformBlockToStructuredBufferTranslation.md
```
Environment:
OS : Windows 11
Browser : Chrome 114
GPU : NVIDIA GeForce RTX 2060 
Backend : ANGLE (Direct3D11 vs_5_0 ps_5_0, D3D11)

**Comment by bejado** [comment]
Interesting, the second link doesn't even load for me and I'm forced to kill the page. I'm on a Macbook Pro. I wonder if https://github.com/google/filament/pull/6848 is related to this at all.

**Reply by bejado** [reply]
I can reproduce with our WebGL samples, I'll try to bisect our commits.

**Reply by bejado** [reply]
Yeah, after doing a bisect, https://github.com/google/filament/pull/6848 seems to be the culprit. I'll need to investigate further.

**Comment by bejado** [comment, marked as answer]
After some investigation, it appears this is due to slow shader compilation on WebGL. Shader compilation is already slow with WebGL due to a [known bug with Chrome](https://github.com/google/filament/issues/6615). However, after https://github.com/google/filament/pull/6848, we're now pre-compiling all the variants of a single material, which exacerbates the problem.

Thankfully the workaround (for now) is straightforward: https://github.com/google/filament/pull/6925
